Mission 2: render, useState 작성 / rerender 작업 필요

youngseo-kangg opened the pull request:
- render 작성
- useState 작성
- rerender 작업중
  - 새로 업데이트 된 useState를 가지고 새로운 DOM을 생성하고, 새로운 DOM을 이전의 DOM과 비교하는 로직을 구현해서 새로 리렌더링 되는 부분이 작동 되어야 하는데, 새로운 state를 어떻게 반영해야 될지부터 감이 안잡히네요,,.....

youngseo-kangg committed:
feat: render 함수 구현 및 관련 타입 수정
 - node name 관련 타입 확장
fix: React.Fragment 에러 관련 수정
 - Fragment 에러 관련 변수 작성,  jsxInject 시 React import 추가

youngseo-kangg committed:
feat: render 내 number 렌더링 관련 조건 추가, useState 작업

youngseo-kangg committed:
feat: useState 값 비교 관련 utils/deepEqual 작성, todo 생성
fix: import jsx-runtime 관련 구조 수정

youngseo-kangg committed:
feat: rerender 작업중

YOOJS1205 commented on src/lib/util/index.ts:
사실 React는 얕은 비교 (shallowEqual)를 활용하여 변경점을 감지합니다.
`deepEqual` 유틸 함수를 잘 작성해주셨지만, `shallowEqual` 함수를 작성하여 적용해주시면 좋을 것 같아요!

보완하시면서 깊은 비교와 얕은 비교의 차이점에 대해서도 학습해보면 좋을 것 같아요!

YOOJS1205 commented on src/lib/react/render.ts:
```diff
@@ -0,0 +1,115 @@
+  render(node(), container);
+};
+
+// !! 작동 X -> 새로운 state 반영 안됨 / diffing, reconciliation 로직 구현 필요
```
사실 diffing 알고리즘 구현은 3주차 과제입니다!
2주차에는 `shallowEqual`을 통해 값이 변경되면 다시 `render` 함수를 호출해주시면 됩니다!

YOOJS1205 commented on src/lib/react/render.ts:
```diff
@@ -0,0 +1,115 @@
+ * @param {IReactNode | string} node - 렌더링할 가상 DOM 노드 또는 문자열
+ * @param {HTMLElement} container - 렌더링 대상이 되는 실제 DOM 요소
+ */
+const render = (node: IReactNode | string, container: HTMLElement) => {
```
`render` 함수를 잘 작성해주셨습니다! 몇 가지 보완되면 좋을 점들을 정리해드릴게요.

* Virtual DOM을 기반으로 실제 DOM을 생성하는 함수와 렌더링하는 함수를 분리해보면 어떨까요? 각각 함수의 역할이 더 명확해질 것 같아요! `createDOMElement`, `render` 함수 요런식으로 분리해보면 좋을 것 같아요. 함수 명은 자유입니다.

* React에서 `null`, `undefined` 타입을 어떻게 처리하는지 확인해보시고, 적용하면 좋을 것 같아요!

YOOJS1205 commented on src/lib/react/render.ts:
```diff
@@ -0,0 +1,115 @@
+};
+
+// !! 작동 X -> 새로운 state 반영 안됨 / diffing, reconciliation 로직 구현 필요
+const rerender = () => {
```
2주차 과제에서는 `render` 함수를 호출하는 정도로 마무리하셔도 좋습니다!

YOOJS1205 commented on src/lib/react/hooks/useState.ts:
```diff
@@ -0,0 +1,30 @@
+
+type SetState<T> = (newState: T | ((prevState: T) => T)) => void;
+const stateStore: unknown[] = []; // 여러 useState 호출을 위한 상태 저장 배열
+let stateIndex = 0; // useState가 여러 개일 때 각 state를 구분하는 인덱스
```
`stateIndex` 변수가 전역에 있어서 보호받지 못하고 있는 상태입니다. 클로저를 사용하여 보호될 필요가 있어 보여요!

구조를 아래와 같이 잡으면 어떨까요?

```ts
const createRenderer = () => {
  // 보호받아야할 변수들 선언 (현재 컴포넌트, root, stateIndex 등)

  const render = () => {...};
  const reRender = () => {...};
  const useState = () => {...};

  return { render, reRender, useState }
}
```

YOOJS1205 commented on src/lib/react/hooks/useState.ts:
```diff
@@ -0,0 +1,30 @@
+
+  const setState: SetState<T> = (newState) => {
+    const prevState = stateStore[currentIndex] as T;
+    const nextState =
```
함수형 업데이트를 지원하게 잘 작성해주셨네요!

함수형 업데이트는 상태를 동기적으로 관리할 때 중요할 역할을 합니다. **useState의 비동기적 특성** 키워드에 대해서 공부해보시면 좋을 것 같아요!

YOOJS1205 reviewed (commented):
3주차 과제 내용까지 고민해주셨네요!
지금 가지고 계신 고민들은 3주차 과제를 하시면서 자연스럽게 해결될 거에요!

현재 `render`, `reRender`, `useState`에서 참조하고 있는 상태들이 전역에 선언되어있어 쉽게 오염될 가능성이 있습니다.

클로저를 사용하여 보호받을 수 있는 구조로 개선되면 좋을 것 같습니다! 고생하셨습니다.

youngseo-kangg committed:
fix: shallowEqual 작성 및 setState 내 deepEqual -> shallowEqual 변경

youngseo-kangg committed:
fix: declare global 관련 eslint 설정 수정
feat: 클로저 관련 useState 위치 수정 / render 내 DOM 생성 기능 createDOMElement로 분리 / util typeCheck 기능 추가
